contracts/DefaultStrategy.sol

```diff
@@ -522,248 +436,288 @@ contract DefaultStrategy is UUPSOwnableUpgradeable, UsingRegistryUpgradeable, Ma
+            revert GroupNotActive(group);
+        }
+        activeGroups.remove(group);
+        unsortedGroups.remove(group);
+
+        uint256 groupTotalStCeloVotes = stCELOInGroup[group];
+
+        if (groupTotalStCeloVotes > 0) {
+            updateGroupStCelo(group, groupTotalStCeloVotes, false);
+            address[] memory fromGroups = new address[](1);
+            uint256[] memory fromVotes = new uint256[](1);
+            fromGroups[0] = group;
+            fromVotes[0] = IManager(manager).toCelo(groupTotalStCeloVotes);
+            (address[] memory toGroups, uint256[] memory toVotes) = _generateVoteDistribution(
+                IManager(manager).toCelo(groupTotalStCeloVotes),
+                false
+            );
+            IManager(manager).scheduleTransferWithinStrategy(
+                fromGroups,
+                toGroups,
+                fromVotes,
+                toVotes
+            );
         }
 
-        GroupWithVotes[] memory sortedGroups;
-        uint256 availableVotes;
-        (sortedGroups, availableVotes) = getSortedGroupsWithVotes(activeGroups.values());
-        if (availableVotes < withdrawal) {
-            revert CantWithdrawAccordingToStrategy();
+        emit GroupRemoved(group);
+    }
+
+    /**
+     * @notice Distributes votes by computing the number of votes each active
+     * group should either receive of should be subtracted from.
+     * @param votesAmount The amount of votes to distribute.
+     * @param withdraw Generation for either desposit or withdrawal.
+     * @return finalGroups The groups that were chosen for distribution.
+     * @return finalVotes The votes of chosen finalGroups.
+     */
+    function _generateVoteDistribution(uint256 votesAmount, bool withdraw)
+        private
+        returns (address[] memory finalGroups, uint256[] memory finalVotes)
+    {
+        if (activeGroups.getNumElements() == 0) {
+            revert NoActiveGroups();
         }
-        availableVotes -= withdrawal;
 
-        uint256 numberGroupsWithdrawn = numberGroups;
-        uint256 targetVotes = availableVotes / numberGroupsWithdrawn;
+        uint256 maxGroupCount = Math.min(
+            withdraw ? maxGroupsToWithdrawFrom : maxGroupsToDistributeTo,
+            activeGroups.getNumElements()
+        );
 
-        for (uint256 i = 0; i < numberGroups; i++) {
-            if (sortedGroups[i].votes <= targetVotes) {
-                numberGroupsWithdrawn--;
-                availableVotes -= sortedGroups[i].votes;
-                targetVotes = availableVotes / numberGroupsWithdrawn;
-            } else {
+        address[] memory groups = new address[](maxGroupCount);
+        uint256[] memory votes = new uint256[](maxGroupCount);
+
+        address votedGroup = withdraw ? activeGroups.getHead() : activeGroups.getTail();
+        uint256 groupsIndex;
+        uint256 groupVotes;
+
+        for (groupsIndex = 0; groupsIndex < maxGroupCount; groupsIndex++) {
```

Comment:
This code is very confusing with all the branching/nesting, difficult to tell if it's correct. Would be cleaner with the attempted sorting helper mentioned above, and handling the deposit/withdrawal cases in separate helpers.

Might even be worth it to just have separate functions for the full withdrawal/deposit algorithms, rather than using a `withdraw` flag. E.g., does `groupVotes` have the same semantics in the two cases below? Seems to mean "total votes currently cast on group" in the withdraw case, and "total castable votes for group" in the deposit case.

Reply:
Refactored